### PR TITLE
Memcached storage provider.

### PR DESCRIPTION
**Pull request by johnou**

### CODE REVIEW

File: actors/core/src/main/java/com/ea/orbit/actors/runtime/OrbitActor.java [hunk 1]
```diff
@@ -120,9 +121,9 @@ protected T state()
      *
      * @return a completion promise
      */
-    protected Task<Boolean> readState()
+    protected Task<Void> readState()
```
Comment (DanielSperry) [inline actors/core/src/main/java/com/ea/orbit/actors/runtime/OrbitActor.java]: Could you document the information conveyed by the Boolean in the return type?

Comment (DanielSperry) [inline actors/core/src/main/java/com/ea/orbit/actors/runtime/OrbitActor.java]: Cool, you removed the unused boolean. I wonder if it shouldn't signal if the data was found or not.

**Comment by quilombodigital** [comment]
Yes, I found it interesting too, but the question is, what the application can do with the boolean besides logging or trying again?

**Comment by quilombodigital** [comment]
Real world scenario, the database goes down... the actors keep running.... the database goes up again, hum... I guess that only actors that were deactivated with a failed writestate will loss data. Can we avoid somehow postpone actors deactivation when there were writestate failures?

**Comment by DanielSperry** [comment]
Yes, but that turns one problem into another: memory.

**Comment by quilombodigital** [comment]
Ok, so make it a choice. The system by default keeps the actor activated until a valid writestate. And the user can add a special provider that can choose in what situations the actors can be deactivated despite the writestate success. Another approach would be to configure a max memory limit for this situation, where it will deactivate the oldest in this situation when a memory limit is reached.

File: actors/providers/memcached/README.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Orbit Actors Storage Provider Memcached
+============
+
+Here be dragons, the Memcached persistence test invokes "flush_all". Running it against a shared Memcache instance will result in data loss.
```
Comment (DanielSperry) [inline actors/providers/memcached/README.md]: scary

**Comment by DanielSperry** [comment]
There was a reason not to accept the provider's data object into the actor.
In the future we'd like to be able to manage the actor state.
That's something that for instance Orleans does.

In another life and company I implemented a transactional system that also monitored all access to the application objects. So this would be a simple thing to do. I'm not sure if it's a good thing, though. Now I come to think of this sort of monitoring as an early optimization bound to be more expensive than just cloning the thing.

So while we decide if the object will or will not be updated in place.
@johnou have you looked into libraries like [Dozer](http://dozer.sourceforge.net/) or [kryo](https://github.com/EsotericSoftware/kryo) they make it possible to clone or copy objects.

What would allow the in place replacement of object would be something like:

```
Task<Boolean> readState(ActorReference<?> reference, Object state) {
    Object o = readFromSomeDatabase(...);
    mapper.copy(o, state);
    return done();
}
```

I understand your concern with asynchronous reads (where the application neglects to `await(readState());` )
However if the application chooses to use async reads, it will be in trouble with the atomic state as well.
Because it will still be possible that two subsequent calls to `state()` will return different objects.
I agree that that is better than the object fields changing mysteriously during the same call.


So as you can see this discussion is not over yet. 
And I'd really appreciate to hear your arguments.

**Comment by johnou** [comment]
@DanielSperry thanks for such a detailed insight of what you might have planned for the future. To be honest I hadn't thought as far as monitoring application objects and I was not happy with changing the API of the storage provider so I will go ahead and revert those changes. 

Dozer looks reasonable in terms of updating the actor state object so I will update the Memcached provider to use that.

I also updated the test to just loop all known keys, inefficient sure, but not dangerous.

File: actors/providers/memcached/src/main/java/com/ea/orbit/actors/providers/memcached/MemCachedStorageProvider.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+ * @author Johno Crawford (johno@sulake.com)
+ */
+public class MemCachedStorageProvider implements IStorageProvider
+{
+
+    private Mapper mapper;
+    private MemCachedClient memCachedClient;
+    private MemCachedStorageHelper memCachedStorageHelper;
+
+    private boolean useShortKeys = false;
```
Comment (johnou) [inline actors/providers/memcached/src/main/java/com/ea/orbit/actors/providers/memcached/MemCachedStorageProvider.java]: Useful for innodb backed memcached storage.

File: actors/providers/memcached/src/main/java/com/ea/orbit/actors/providers/memcached/ByteUtils.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            {
+                enableResolveObject(true);
+            }
+            @Override
+            protected Object resolveObject(final Object obj) throws IOException
+            {
+                if (obj instanceof ReferenceReplacement)
+                {
+                    ReferenceReplacement replacement = (ReferenceReplacement) obj;
+                    // TODO: help?
```
Comment (johnou) [inline actors/providers/memcached/src/main/java/com/ea/orbit/actors/providers/memcached/ByteUtils.java]: @DanielSperry any suggestions on what we can do from this context?

File: actors/providers/postgresql/src/main/java/com/ea/orbit/actors/providers/postgresql/PostgreSQLStorageProvider.java [hunk 5]
```diff
@@ -55,7 +55,7 @@
     private String username;
     private String password;
 
-    private Connection conn;
+    private Connection connection;
     private PreparedStatement insertState;
```
Comment (johnou) [inline actors/providers/postgresql/src/main/java/com/ea/orbit/actors/providers/postgresql/PostgreSQLStorageProvider.java]: general observation of the postgres implementation, I would consider removing the sync keywords and creating a new prepared statement every time we need to read / write and close it in a finally block..

File: actors/providers/memcached/src/main/java/com/ea/orbit/actors/providers/memcached/ByteUtils.java (hunk 4 above)
Comment (DanielSperry) [inline actors/providers/memcached/src/main/java/com/ea/orbit/actors/providers/memcached/ByteUtils.java]: `IActor.getReference` if the class is an IActor.
else `IActorObserver.getReference()` which doesn't exist yet. I'll add it shortly.

**Comment by DanielSperry** [comment]
@johnou I've added two utility methods for restoring references:  https://github.com/electronicarts/orbit/commit/18c036a62347e7434d74b676d4e489af99f49676

**Comment by JoeHegarty** [comment]
Merged, thanks Johno.

**Comment by DanielSperry** [comment]
:+1:

**Pull request merged**